Feature: add Drizzle support @effect/sql-drizzle

ecyrbe opened the pull request:
<!--
Before submitting a Pull Request, please ensure you've done the following:

- 📖 Read our Contributing Guide: https://github.com/effect-ts/.github/blob/main/CONTRIBUTING.md
- 📖 Read our Code of Conduct: https://github.com/effect-ts/.github/blob/main/CODE_OF_CONDUCT.md
- 👷‍♀️ Create small PRs. In most cases this will be possible.
- 📝 Use descriptive commit messages.
- ✅ Provide tests for your changes if applicable.
- 📗 If your change requires documentation, please update the relevant documentation.
- 📝 Create a changeset for your changes. This helps in tracking and communicating the changes effectively.
- ⏳ Please be patient! We will do our best to review your pull request as soon as possible.

NOTE: Pull Requests from forked repositories will need to be reviewed by a team member before any CI builds will run.
-->

## Type

<!--
What type of change is this? Please check all applicable.
-->

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

<!--
Please add a brief summary/description of the pull request here.
-->
This PR add support for drizzle in effect.
It uses internal Drizzle DB, and disable drizzle runtime execution (execute and transaction) to leverage on effect sql client

## example
```ts
const SqlLive = Sqlite.client.layer({
  filename: Config.succeed("test.db")
})
const DrizzleLive = SqliteDrizzle.layer.pipe(
  Layer.provide(SqlLive)
)
const DatabaseLive = Layer.mergeAll(SqlLive, DrizzleLive)

// usage

const users = D.sqliteTable("users", {
  id: D.integer("id").primaryKey(),
  name: D.text("name")
})

Effect.gen(function*() {
  const sql = yield* Sql.client.Client
  const db = yield* SqliteDrizzle.SqliteDrizzle
  yield* sql`CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT)`
  const inserted = yield* db.insert(users).values({ name: "Alice" }).returning()
  yield* Console.log(inserted)
  const selected = yield* db.select().from(users)
  yield* Console.log(selected)
  const updated = yield* db.update(users).set({ name: "Bob" }).where(gte(users.id, 1)).returning()
  yield* Console.log(updated)
  const deleted = yield* db.delete(users).where(gte(users.id, 1)).returning()
  yield* Console.log(deleted)
}).pipe(
  Effect.provide(DatabaseLive),
  Effect.runPromise
)
```
## Related

<!--
For pull requests that relate or close an issue, please include them below. We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull request to issue 1234 and automatically
close the issue once we merge the pull request.
-->

- Related Issue #
- Closes #

ecyrbe commented on packages/sql-drizzle/test/pg.test.ts:
```diff
@@ -0,0 +1,53 @@
+  let makeClient: Effect.Effect<Pg.client.PgClient, never, Scope>
+
+  beforeAll(async () => {
+    pgContainer = await new PostgreSqlContainer("postgres:alpine").start()
```
using testcontainer, not sure that's the standard way to test db in effect. tell me if it's ok.

mikearnaldi replied:
testcontainer is really great, I usually wrap it in a layer so I can have a TestDb that spins up and down automatically, but this looks fine too

ecyrbe commented on packages/sql-drizzle/src/pg.ts:
```diff
@@ -0,0 +1,58 @@
+  const client = yield* Client.Client
+  // instanciate the db without a client, since we are going to attach the client to the QueryBuilder
+  const db = drizzle({
+    options: {
+      parsers: [],
+      serializers: []
+    }
```
This is to make drizzle patching of postgres-js client instance working. drizzle patching is not used in our case, since db instance is just used to convert query builder postgres dialect to sql

mikearnaldi commented:
```
packages/sql-drizzle/src/internal/registry.ts
```

Seems like the above file creates a global map Dialect<->Client but the Client is scoped and we might have many different clients in the same project.

Additionally, patching the global drizzle type would make every client effectable, even if registration didn't happen.

I think a better idea would be to yes make every drizzle query effectable but use `Effect<T, SqlError, Client>` so that the client is passed to the query when used

ecyrbe commented:
> ```
> packages/sql-drizzle/src/internal/registry.ts
> ```
> 
> Seems like the above file creates a global map Dialect<->Client but the Client is scoped and we might have many different clients in the same project.
> 

This is explained in the comment of the code. We store the `instance` of the dialect, and a new instance of the dialect is created for each db... so each client will have a different dialect associated with it, effectively allowing to have many clients in the same project.

> Additionally, patching the global drizzle type would make every client effectable, even if registration didn't happen.
> 
> I think a better idea would be to yes make every drizzle query effectable but use `Effect<T, SqlError, Client>` so that the client is passed to the query when used

Nice idea, it would indeed remove this global registry. i'll try it to see if it works and report back ASAP.

mikearnaldi commented:
> > ```
> > packages/sql-drizzle/src/internal/registry.ts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Seems like the above file creates a global map Dialect<->Client but the Client is scoped and we might have many different clients in the same project.
> 
> This is explained in the comment of the code. We store the `instance` of the dialect, and a new instance of the dialect is created for each db... so each client will have a different dialect associated with it, effectively allowing to have many clients in the same project.
> 
> > Additionally, patching the global drizzle type would make every client effectable, even if registration didn't happen.
> > I think a better idea would be to yes make every drizzle query effectable but use `Effect<T, SqlError, Client>` so that the client is passed to the query when used
> 
> Nice idea, it would indeed remove this global registry. i'll try it to see if it works and report back ASAP.

Ok so it isn't the dialect itself but a specific instance created with it's own client, then I think it could work, it might even be better compared to `Effect<T, SqlError, Client>` as it would capture the client on creation but the issue remains on having all db instances effectable regardless of the client being attached.

Also small note, `packages/sql-drizzle/src/internal/drizzle-patch.types.ts` should not be on `/internal` given that it is exposed

ecyrbe commented:
> Ok so it isn't the dialect itself but a specific instance created with it's own client, then I think it could work, it might even be better compared to `Effect<T, SqlError, Client>` as it would capture the client on creation but the issue remains on having all db instances effectable regardless of the client being attached.

I have a solution for that, patch SQLiteSelectBase, SQLiteInsertBase directly. that's what i did first. but the types are ugly :)
It's less pretty to maintain, but i can do it.

> Also small note, `packages/sql-drizzle/src/internal/drizzle-patch.types.ts` should not be on `/internal` given that it is exposed

Indeed.

tim-smart commented on packages/sql-drizzle/src/internal/drizzle-patch.types.ts:
```diff
@@ -0,0 +1,10 @@
+/**
+ * QueryPromise class is used for every Drizzle QueryBuilders, it's what allows Builders to be awaitable
+ * So since we need all QueryBuilders to also be effectable, we patch it's interface to extend Effect
+ * We don't however monkey patch the QueryPromise class itself, but the QueryBuilders that extend it to be able to attach the right client to them
```
Patching QueryPromise might be cleaner. Is it exposed?

ecyrbe replied:
yes, it's exposed

ecyrbe replied:
But patching QueryPromise with dependency injecting the client will patch all instances (not just the one we declared), it's indeed doable, but @mikearnaldi has concerns about patching globally the types, so patching the propotype globally would be even more intrusive.

mikearnaldi replied:
> But patching QueryPromise with dependency injecting the client will patch all instances (not just for the client we declared), it's indeed doable, but @mikearnaldi has concerns about patching globally the types, so patching the propotype globally would be even more intrusive.

I don't really have concerns if the patching is valid, the issue would be patching instances globally that can't be used in an effect environment, if the patch adds `Effect<X, SqlError, Client>` it is valid because the client will need to be provided so as long as augmentation and patching are done at the same time it should work

ecyrbe replied:
So i tested it, and patching QueryPromise does not work. No idea why, i get this error :
```md
Unknown Error: TypeError: yield* (intermediate value)(intermediate value)(intermediate value)(intermediate value)(intermediate value)(intermediate value) is not iterable
 ❯ test/sqlite.test.ts:31:43
     29|       const inserted = yield* db.insert(users).values({ name: "Alice" }).returning()
     30|       assert.deepStrictEqual(inserted, [{ id: 1, name: "Alice" }])
     31|       const selected = yield* db.select().from(users)
       |                                           ^
```
have you ever seen it ? If no solution, i'll keep the patching at builder level.

tim-smart replied:
I have another idea, will try it out and report back if it works.

tim-smart replied:
Updated #2860

ecyrbe commented:
So, i updated the PR.
It now uses dependency injection for client instead of registry. I feel it's safer this way.
It allows to no have any dependency on private `dialect` property, making this implementation more stable in time
I added the type patching at Builder level instead of `QueryPromise`. Like i sead it's ugly, and risky, since any parameter change in the builder types (renaming, adding new ones) will break typescript inference for users.

I feal that the best combination is to :
- patch QueryBuilders without registry and use dependency injection
- patch QueryPromise type as it only have one type parameter making it more stable

what do you think ? do i revert back to QueryPromise type Patching ?

ecyrbe commented:
closing, since tim re-took the lead on this.
